GH-35627: [Format][Integration] Add string-view to arrow format (#37526)
apache/arrow: docs/source/format/Columnar.rst

* **Primitive (fixed-size)**: a sequence of values each having the
  same byte or bit width
* **Variable-size Binary**: a sequence of values each having a variable
  byte length. Three variants of this layout are supported using
  * 32-bit offset encoding
  * 64-bit offset encoding

>   * 128-bit view-or-inline encoding

Reviewer: I can see both sides of the argument. From the standpoint of "it encodes the same type of data" (a variable-size binary value) then including it in this list makes sense, and is perhaps somewhat less confusing for a reader of the specification.

>   * 128-bit view-or-inline encoding

Reviewer 2: As in: a separate bullet from `Variable-size Binary`? If so I'd agree, though if we do we should make sure to differentiate it by saying something like "Unlike the Variable-size binary layout, the elements in this layout may be out-of-order or even overlapping." Without any clarification on the differences, I agree it might be confusing for a reader of the spec.

>   * 128-bit view-or-inline encoding

Author: I'll amend the list with a new bullet point for `Views of Variable-size Binary`

    |----------------|-----------------------|
    | joemark        | unspecified (padding) |

.. versionadded:: Arrow Columnar Format 1.4

> Variable-size Binary View Layout

Reviewer 3: ... and indeed here this is described as a different layout.

Reviewer: I would agree that this should at least be made consistent with the decision made above (perhaps just added as a subsection within Variable-size Binary).

> .. versionadded:: Arrow Columnar Format 1.4
> 
> Variable-size Binary View Layout
> --------------------------------

Reviewer 3: Should move the `versionadded` below the title to avoid any confusion
```suggestion
Variable-size Binary View Layout
--------------------------------

.. versionadded:: Arrow Columnar Format 1.4
```

> Each value in this layout consists of 0 or more bytes. These characters'

Reviewer 3: ```suggestion
Each value in this layout consists of 0 or more bytes. These bytes'
```

locations are indicated using a **views** buffer, which may point to one
of potentially several **data** buffers or may contain the characters
inline.

The views buffer contains `length` view structures with the following layout:

::

    * Short strings, length <= 12
      | Bytes 0-3  | Bytes 4-15                            |
      |------------|---------------------------------------|
      | length     | data (padded with 0)                  |

    * Long strings, length > 12
      | Bytes 0-3  | Bytes 4-7  | Bytes 8-11 | Bytes 12-15 |
      |------------|------------|------------|-------------|
      | length     | prefix     | buf. index | offset      |

In both the long and short string cases, the first four bytes encode the
length of the string and can be used to determine how the rest of the view
should be interpreted.

In the short string case the string's bytes are inlined- stored inside the
view itself, in the twelve bytes which follow the length.

> In the long string case, a buffer index indicates which character buffer
> stores the characters and an offset indicates where in that buffer the
> characters begin. Buffer index 0 refers to the first character buffer, IE

Reviewer 3: ```suggestion
In the long string case, a buffer index indicates which data buffer
stores the data bytes and an offset indicates where in that buffer the
data bytes begin. Buffer index 0 refers to the first data buffer, IE
```

the first buffer **after** the validity buffer and the views buffer.
The half-open range ``[offset, offset + length)`` must be entirely contained
within the indicated buffer. A copy of the first four bytes of the string is
stored inline in the prefix, after the length. This prefix enables a
profitable fast path for string comparisons, which are frequently determined
within the first four bytes.

Views must be aligned to an 8-byte boundary. This restriction enables more
efficient interoperation with systems where the index and offset are replaced

> by a raw pointer. All integers (length, buffer index, and offset) are unsigned

Reviewer 4: I have to say I don't really understand this requirement, if the data is being mutated that necessarily implies a copy anyway?

Author: When swizzling between index/offset and pointers, we frequently have sole ownership of the views buffer and can overwrite it directly. This isn't zero cost but it does save an allocation. This optimization isn't possible if an index/offset is not aligned to an 8 byte boundary since we can't overwrite them with a pointer in that case.

Reviewer 4: Can't we just make the swizzling predicate - exclusive and aligned? It seems odd to mandate this and not for other types?

> by a raw pointer. All integers (length, buffer index, and offset) are unsigned

Reviewer 3: 1. Other types don't have a specific alignment requirement even where it might be beneficial. Instead there is a general "Buffer Alignment and Padding" section. I don't think binary views are not special enough to warrant this.
2. I don't know of any CPU feature that would make alignment of pointer data more imperative than, say, alignment of 64-bit data.
3. The spec certainly shouldn't appear to condone or promote non-Arrow formats disguised as Arrow ("systems where the index and offset are replaced by a raw pointer").

Reviewer 3: I don't understand what "swizzle" means in this context, but I still don't understand how this is different from, say, MONTH_DAY_NANO_INTERVAL.

> by a raw pointer. All integers (length, buffer index, and offset) are unsigned

Reviewer 3: Please let's please not make a special case for binary-view here. We use signed offsets and lengths everywhere else.
```suggestion
by a raw pointer. All integers (length, buffer index, and offset) are signed
```

Author: We should discuss this on the ML, I'll raise this there

Author: https://lists.apache.org/thread/x3mh2zh5qlm3fss2716yn0dkortz6z3w

Author: It's not a requirement to be imposed on sizes in all cases going forward. It's only relevant to Utf8View because existing implementations of string view already use unsigned integers here, so the concern is whether to keep compatibility with those or maintain the convention of the arrow format. If you'd like to +1 signed integers, please say so on the ML!

Author: Per the mailing list, I've updated this to signed integers.

      | length     | prefix     | buf. index | offset      |

In both the long and short string cases, the first four bytes encode the
length of the string and can be used to determine how the rest of the view
should be interpreted.

In the short string case the string's bytes are inlined- stored inside the
view itself, in the twelve bytes which follow the length.

In the long string case, a buffer index indicates which data buffer
stores the data bytes and an offset indicates where in that buffer the
data bytes begin. Buffer index 0 refers to the first data buffer, IE
the first buffer **after** the validity buffer and the views buffer.
The half-open range ``[offset, offset + length)`` must be entirely contained
within the indicated buffer. A copy of the first four bytes of the string is
stored inline in the prefix, after the length. This prefix enables a
profitable fast path for string comparisons, which are frequently determined
within the first four bytes.

All integers (length, buffer index, and offset) are unsigned
for compatibility with engines which already implement these views.

This layout is adapted from TU Munich's `UmbraDB`_.

> It is worth noting that maintaining recommended `Buffer Alignment and Padding`_

Reviewer 3: I don't think it is useful to keep this paragraph. It does not make it easier to understand or implement the Arrow format. Ease of conversion with other formats is not the purpose of this document.

Reviewer 3: However, you may perhaps state that the recommended alignment is 16 bytes (or is it 8?) for ease of implementation reasons.

Author: In light of @tustvold's and your objections to the note, I'll remove it from Columnar.rst. It can be part of the doccomment for the swizzling function

of memory buffers for each layout.

.. csv-table:: Buffer Layouts
   :header: "Layout Type", "Buffer 0", "Buffer 1", "Buffer 2", "Variadic Buffers"
   :widths: 30, 20, 20, 20, 20

   "Primitive",validity,data,,
   "Variable Binary",validity,offsets,data,

>    "Variable Binary - View",validity,packed headers,,data

Reviewer 3: Nit, but I don't think the hyphen or dash is useful here?
Also, we use the term "view" rather than "headers" above and we don't mention packing.
```suggestion
   "Variable Binary View",validity,views,,data
```
